### ISSUE TITLE
Different JSON structure and details when checking from web or command line

### ISSUE THREAD

**Issue by jaimeiniesta** [validator/validator#53]
When checking the same page, first on the web interface, and then from the command line, I'm getting a different JSON structure and different details. The issues reported are the same, but the web version includes additional details like the extract and hilite:

https://gist.github.com/jaimeiniesta/4f8bc92c00e8341002b8

I'm using the latest validator release (16 February 2015), on OSX Yosemite with this Java version:

    java version "1.6.0_65"
    Java(TM) SE Runtime Environment (build 1.6.0_65-b14-466.1-11M4716)
    Java HotSpot(TM) 64-Bit Server VM (build 20.65-b04-466.1, mixed mode)

Is this difference intentional? Is there a way to get the same JSON structure on both, or at least to have as well the extra details that the web version has?

Can this be related with the Java version I'm using and if so, which one is the recommended to use validator?

Thank you!

**Comment by sideshowbarker** [comment]
Interesting. It’s definitely not intentional, and it almost certainly has nothing to do with your client java version.

But that said, in general for building and running the validator I'd recommend using at least Java7 (1.7) if not Java8. Myself I do most of my development on OSX Yosemite, using Java8. I don't plan on intentionally making any changes that require Java7 or 8, but the thing is, Java6 is now very very old, and has been end-of-lifed by Oracle, so going forward, ensuring continued compatibility with Java6  is not a goal for the validator project.

**Comment by sideshowbarker** [comment]
Anyway, will try to make some time to investigate this today

**Comment by jaimeiniesta** [comment]
Thanks, I've updated to Java 1.8 and I can confirm this still happens. I'm running the checker like this:

```bash
java -jar vnu.jar --format json http://validationhell.com
```

I get the same results if I pass it a local HTML file.

**Comment by jaimeiniesta** [comment]
I've tried this on Ubuntu Precise 32 and I also get a different JSON structure on the command line version.

**Comment by sideshowbarker** [comment]
Please test this change using the latest nightly:
https://sideshowbarker.net/nightlies/vnu.jar

**Comment by jaimeiniesta** [comment]
Thank you, I just compared the JSON produced by the latest nightly on CLI, VS the JSON produced currently on http://validator.w3.org/nu and it's almost the same structure now.

There's just one thing that is still different as you can see [in this gist](https://gist.github.com/jaimeiniesta/3fb78fb04d645f5b0ef3), the URL of the checked document is shown in the top level when checking from web:

```json
{
  "url": "http://validationhell.com",
  "messages": [{
    "type": "info",
    "message": "The Content-Type was “text/html”. Using the HTML parser."
  }, {
    "type": "info",
    "message": "Using the schema for HTML with SVG 1.1, MathML 3.0, RDFa 1.1, and ITS 2.0 support."
  }, {
    "type": "error",
    "lastLine": 56,
    "firstLine": 55,
    "lastColumn": 64,
    "firstColumn": 37,
    "message": "The “align” attribute on the “img” element is obsolete. Use CSS instead.",
    "extract": " href=\"/\"><img\n src=\"/images/fire.png\" align=\"absmiddle\" width=\"30\" hspace=\"5\"><stron",
    "hiliteStart": 10,
    "hiliteLength": 69
  }, 
```

but when checking on CLI, the URL is included on each message:

```json
{
    "messages": [
        {
            "type": "info",
            "url": "http://validationhell.com",
            "lastLine": 1,
            "lastColumn": 109,
            "firstColumn": 1,
            "subType": "warning",
            "message": "Obsolete doctype. Expected “<!DOCTYPE html>”.",
            "extract": "<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Strict//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd\">\n<html",
            "hiliteStart": 0,
            "hiliteLength": 109
        },
        {
            "type": "error",
            "url": "http://validationhell.com",
            "lastLine": 56,
            "firstLine": 55,
            "lastColumn": 64,
            "firstColumn": 37,
            "message": "The “align” attribute on the “img” element is obsolete. Use CSS instead.",
            "extract": " href=\"/\"><img\n src=\"/images/fire.png\" align=\"absmiddle\" width=\"30\" hspace=\"5\"><stron",
            "hiliteStart": 10,
            "hiliteLength": 69
        }
```

I'd go with the version that includes the URL on every message, it makes sense to have it along with the coordinates of the issues found, and if in the future the validator starts checking linked documents (like the CSS validator does), we'll need to specify the URL for each issue.

**Comment by sideshowbarker** [comment]
> Thank you, I just compared the JSON produced by the latest nightly on CLI, VS the JSON produced currently on http://validator.w3.org/nu and it's almost the same structure now.
> There's just one thing that is still different as you can see in this gist, the URL of the checked document is shown in the top level when checking from web: …
> but when checking on CLI, the URL is included on each message:

That difference is actually by design. When you're checking documents using the checker service from http://validator.w3.org/nu or wherever, you're only checking one document during each transaction, so it's not necessary or useful to repeat the URL in each message.

But the CLI is capable for checking multiple documents during a single transaction, so in that case for the GNU output format it's actually necessary to report the URL as part of each message. Otherwise you can't tell which document to associate the message with.

Anyway I'm not inclined to change the output from the Web service to match the CLI output just for the sake of consistency across different contexts. They are two different use cases really.

> if in the future the validator starts checking linked documents (like the CSS validator does), we'll need to specify the URL for each issue

If we ever do add to the HTML checker the ability to check linked documents the way the CSS validator does, then at that time we can change the backend to add the URL to the messages.

**Comment by jaimeiniesta** [comment]
:+1:, totally agree

**Comment by sideshowbarker** [comment]
> > Thank you, I just compared the JSON produced by the latest nightly on CLI, VS the JSON produced currently on http://validator.w3.org/nu and it's almost the same structure now.
> > There's just one thing that is still different as you can see in this gist, the URL of the checked document is shown in the top level when checking from web: …
> > but when checking on CLI, the URL is included on each message:
> That difference is actually by design. When you're checking documents using the checker service from http://validator.w3.org/nu or wherever, you're only checking one document during each transaction, so it's not necessary or useful to repeat the URL in each message.

So, thinking about this more, the assumption “When you're checking documents using the checker service from http://validator.w3.org/nu or wherever, you're only checking one document during each transaction” doesn't necessarily always hold true. While it is true when using the (current) Web frontend, it’s not always true of course when you’re using the checker’s HTTP/Web-service API rather than the frontend—because in that case if your app is looping through a list of files and sending separate requests for each, it can be very useful to have each message include a filename/URL.

And while it’s true that your app code could keep its own record of each filename/URL, and rely on that in its output to identify the file to the user, you shouldn’t need for your app to need to do that if you with have the filename/URL shown in each message and can just pass that on as-is to the user.

> Anyway I'm not inclined to change the output from the Web service to match the CLI output just for the sake of consistency across different contexts. They are two different use cases really.

I realize now that they are not entirely different use cases—at least the use cases for the HTTP/Web-services API can be very similar to use cases for the CLI.

So I reckon I'm going to change this, actually. Lemme know if you think that’s a bad idea.

One thing is, even after I make this change, the JSON output and the XML output are still going to also show a URL at the top level of the message output—along with showing a URL for each message. That’s redundant but dropping it from the top level now would break backward compatibility for any apps or third-party tools that are currently relying on the URL being emitted at the top level.

**Comment by jaimeiniesta** [comment]
I don't know, I think that adding the URL to every message doesn't add any
value to the output from the web UI currently and if it's still going to be
available at the top, it will just make the JSON weigh a bit more.

It would be different if the UI allowed to check multiple documents, in
that case that would be needed.

I just found it a bit weird that the output was different, as I just check
one single document, so I expected the same output.

Another idea: if you just check one document, have the URL returned at the
top level. If you check multiple docs, include their URLs on every message.

In other words, make that depend on what you're requesting, not from where.

**Comment by sideshowbarker** [comment]
> Another idea: if you just check one document, have the URL returned at the top level. If you check multiple docs, include their URLs on every message. In other words, make that depend on what you're requesting, not from where.

There's no way for the servlet code to know whether you're checking one document or many. It’s stateless. All it does is see a request come in, and send a response—just like a Web server.

> I don't know, I think that adding the URL to every message doesn't add any value to the output from the web UI currently and if it's still going to be available at the top, it will just make the JSON weigh a bit more.

True. So I suppose it’s possible to make the behavior be that if you're using the Web frontend, it doesn't add the URL to each message. Does that sound like a good way to resolve this?

**Comment by jaimeiniesta** [comment]
>True. So I suppose it’s possible to make the behavior be that if you're using the Web frontend, it doesn't >add the URL to each message. Does that sound like a good way to resolve this?

I think I'm a bit lost here. Isn't that the same as leaving things as are now?

**Comment by jaimeiniesta** [comment]
> There's no way for the servlet code to know whether you're checking one document or many. It’s stateless. All it does is see a request come in, and send a response—just like a Web server.

I don't know about the implementation of vnu, what I mean is that using the CLI, if we're just checking one document like that:

    java -jar vnu.jar --format json http://example1.com

Then it would make sense to have the URL just being included at the top level, and not on every message, but if we're checking multiple documents:

    java -jar vnu.jar --format json http://example1.com http://example2.com

Then it would make sense to do the opposite, just include the URLs on the messages, and not on the top level.

Also, if from the CLI we're currently able to check multiple documents, I don't see why we wouldn't be able to do so from the web interface. Let's suppose we add a feature to do batch-checking of several documents from the web interface, in that case we'd be on the second example.

But, again, I don't know about the implementation of vnu, for me it's OK to leave it like it is now, it's just that it looked a bit inconsistent from my point of view.

**Comment by sideshowbarker** [comment]
> > True. So I suppose it’s possible to make the behavior be that if you're using the Web frontend, it doesn't >add the URL to each message. Does that sound like a good way to resolve this?
> 
> I think I'm a bit lost here. Isn't that the same as leaving things as are now?

Yeah—sorry for being unclear. What I had meant was that we could keep the behavior the same if you're using the Web frontend, but change it to be different if you're using the Web-services API directly, instead of going through the Web frontend.

But after considering it more, I realize there's really nothing that's broken here that needs fixing, so I'm back to thinking we should just keep everything the way it is.

**Comment by jaimeiniesta** [comment]
I agree.

I was always referering to the Web service API, vs the CLI.

As the web API only accepts one URL, but the CLI accepts multiple URLs, it makes sense that the output structure is different. If some day we change that, so that the web API can accept multiple URLs, it will be the moment to consider this change.

Thanks for your explanations!
